### PR TITLE
add `cupy.setxor1d` api

### PR DESCRIPTION
**Pull request by pri1311**
Linked Issue: #6078 

Hey! I have implememeted numpy.setxor1d but seems like when you provide a multi-dim array with assume_unique as True, numpy raises an error, since their code concatenates array as - np.concatenate((ar1,ar2)) and not np.concatenate((ar1,ar2), axis=None).
However, when assume_unique is false, the array is flattened as it uses the numpy.unique function which returns a flattened array. Do we replicate the same behaviour in CuPy or allow multi-dim arrays even when assume_unique is set to true?

Attaching a screenshot below to explain the above:
159218884-0c0ac806-b821-477b-b68a-bec7ef48077e

Open discussion in numpy - https://github.com/numpy/numpy/issues/14670
Currently, the code I have written does not throw errors for multi-dim arrays, regardless of where `assume_unique` is True or False. However, tests currently don't cover that, since I can't compare it to numpy results. Should I add some manual-comparison tests? Please let me know your opinion.

### CODE REVIEW

**Comment by emcastillo** [comment]
We align to the latest supported numpy version in the master branch. If NumPy errors, CuPy should error too unless there is a good reason that prevents this (device synchronization, checks using divergent threads, etc). In such case it should be clearly stated in the docstring.

**Comment by pri1311** [comment]
> We align to the latest supported numpy version in the master branch. If NumPy errors, CuPy should error too

Alright, thanks. Will make the required changes.

**Comment by asi1024** [comment]
An algorithm using `in1d` may be faster. Could you compare the performance?

**Comment by asi1024** [comment]
@pri1311 Could you try `cupyx.profiler.benchmark`?

File: cupy/_logic/truth.py [hunk 1]
```diff
@@ -255,6 +255,44 @@ def setdiff1d(ar1, ar2, assume_unique=False):
...
+
+    aux = cupy.concatenate((ar1, ar2), axis=None)
+    if aux.size == 0:
+        return aux
+
+    aux.sort()
+    flag = aux
+    flag = cupy.concatenate((cupy.asanyarray([True]), cupy.asanyarray(
+        aux[1:] != aux[:-1]), cupy.asanyarray([True])), axis=None)
+    return aux[flag[1:] & flag[:-1]]
```
Comment (asi1024) [inline cupy/_logic/truth.py:293]: By using `ElementwiseKernel`, we can compute `flag[1:] & flag[:-1]` without allocating additional memories for temporary values.

**Comment by pri1311** [comment]
> @pri1311 Could you try `cupyx.profiler.benchmark`?

Aah yes, on it.

**Comment by pri1311** [comment]
image
These are the current stats, for array sizes
x: (3,2) y: (2,5)
x: (30,20) y: (20,50)
x: (300,200) y: (200,500)

File: cupy/_logic/truth.py (hunk 1 above)
Comment (asi1024) [reply cupy/_logic/truth.py:293]: @pri1311 How about this comment?

Comment (pri1311) [reply cupy/_logic/truth.py:293]: Hey @asi1024! Sorry for the delay, was in between my exams, will get this done asap.

**Comment by pri1311** [comment]
I replaced the following two lines
```
flag = cupy.concatenate((cupy.asanyarray([True]), cupy.asanyarray(aux[1:] != aux[:-1]), cupy.asanyarray([True])))
return aux[flag[1:] & flag[:-1]]
```
with
```
flag = cupy.concatenate((cupy.asanyarray([True]), boolcompare(aux[1:], aux[:-1]), cupy.asanyarray([True])))
return aux[xorkernel(flag[1:],flag[:-1])]
```

```
boolcompare = cupy.ElementwiseKernel(
    'T x, T y',
    'bool z',
    'z = x != y',
    'boolcompare' 
)
```
and this is the change in performance. 
image

I'll try to make further changes so as to avoid using the variable flag altogether by tonight. But would like to know if this is better than before or I could've done better?

Array sizes used for comparison:
```
x = cupy.random.rand(300,200)
y = cupy.random.rand(200,500)
```
Apologies for the delay.
Thank you!

**Comment by pri1311** [comment]
Yes working on that

**Comment by pri1311** [comment]
```
def setxor1dusingkernel(ar1, ar2, assume_unique=False):
    if not assume_unique:
        ar1 = cupy.unique(ar1)
        ar2 = cupy.unique(ar2)

    aux = cupy.concatenate((ar1, ar2), axis=None)
    if aux.size == 0:
        return aux

    aux.sort()

    setxorkernel = cupy.ElementwiseKernel(
        'raw T X, int32 len',
        'T z',
        '''
            if (i == 0) {
                z = (X[0] != X[1]);
            }
            else if (i == len - 1) {
                z = (X[i] != X[i-1]); 
            }
            else {
                z = X[i] != X[i-1] && X[i] != X[i+1];
            }
        ''',
        'setxorkernel' 
    )
    return setxorkernel(aux, aux.size, aux)
```
How about this?
Results:::
image

For array sizes
```
x = cupy.random.rand(300,200)
y = cupy.random.rand(200,500)
```
```
x = cupy.random.rand(3000,2000)
y = cupy.random.rand(2000,5000)
```

**Comment by pri1311** [comment]
Hey, a few tests are failing (and I understand why) let me just make that correction, and will commit the changes. For some reason, the bug didn't show up when I was manually running the function from a separate file.

**Comment by pri1311** [comment]
Hey @asi1024, apologies for the delay, had my final exams going on. Have added the final changes from my side. Can you review them once?

File: cupy/_logic/truth.py [hunk 2]
```diff
@@ -255,6 +255,60 @@ def setdiff1d(ar1, ar2, assume_unique=False):
...
+        ar2 = cupy.unique(ar2)
+
+    aux = cupy.concatenate((ar1, ar2), axis=None)
+    if aux.size == 0:
+        return aux
+
+    aux.sort()
+
+    setxorkernel = cupy.ElementwiseKernel(
+        'raw T X, int32 len',
```
Comment (asi1024) [inline cupy/_logic/truth.py:292]:
```suggestion
        'raw T X, int64 len',
```

File: cupy/_logic/truth.py [hunk 3]
```diff
@@ -255,6 +255,60 @@ def setdiff1d(ar1, ar2, assume_unique=False):
...
+        '''
+            if (i == 0) {
+                z = (X[0] != X[1]);
+            }
+            else if (i == len - 1) {
+                z = (X[i] != X[i-1]);
+            }
+            else {
+                z = X[i] != X[i-1] && X[i] != X[i+1];
+            }
```
Comment (asi1024) [inline cupy/_logic/truth.py:303]: How about simplifying as follows?
```suggestion
        z = (i == 0 || X[i] != X[i-1]) && (i == len - 1 || X[i] != X[i+1])
```

File: cupy/_logic/truth.py [hunk 4]
```diff
@@ -255,6 +255,60 @@ def setdiff1d(ar1, ar2, assume_unique=False):
...
+        ''',
+        'setxorkernel'
+    )
+
+    return aux[setxorkernel(aux, aux.size,
+                            cupy.zeros(aux.size, dtype=cupy.bool8))]
```
Comment (asi1024) [inline cupy/_logic/truth.py:309]:
```suggestion
                            cupy.zeros(aux.size, dtype=cupy.bool_))]
```

File: cupy/_logic/truth.py [hunk 5]
```diff
@@ -255,6 +255,60 @@ def setdiff1d(ar1, ar2, assume_unique=False):
...
+        ar1 = cupy.unique(ar1)
+        ar2 = cupy.unique(ar2)
+
+    aux = cupy.concatenate((ar1, ar2), axis=None)
+    if aux.size == 0:
+        return aux
+
+    aux.sort()
+
+    setxorkernel = cupy.ElementwiseKernel(
```
Comment (asi1024) [inline cupy/_logic/truth.py:291]: Could you define this kernel outside the scope of this function?

File: cupy/_logic/truth.py [hunk 6]
```diff
@@ -4,6 +4,14 @@
 from cupy import _util
 
 
+setxorkernel = cupy._core.ElementwiseKernel(
```
Comment (asi1024) [inline cupy/_logic/truth.py:7]: Could you prefix the kernel name with an underscore?

```suggestion
_setxorkernel = cupy._core.ElementwiseKernel(
```

**Review by asi1024 — APPROVED** [review]

**Comment by asi1024** [comment]
LGTM! Thank you for your PR!

**Pull request merged**